codec-compression/src/main/java/io/netty/handler/codec/compression/BrotliDecompressor.java

```diff
@@ -0,0 +1,150 @@
+                case NEEDS_MORE_INPUT:
+                    if (unusedInput == null) {
+                        return Status.NEED_INPUT;
+                    }
+                    addSomeInput(unusedInput);
+                    if (!unusedInput.isReadable()) {
+                        unusedInput.release();
+                        unusedInput = null;
+                    }
+                    break;
+                case OK:
+                    decoder.push(0);
+                    break;
+                case NEEDS_MORE_OUTPUT:
+                    return Status.NEED_OUTPUT;
+                default:
+                    throw new AssertionError("Unknown status: " + decoder.getStatus());
+            }
+        }
+    }
+
+    @Override
+    public void addInput(ByteBuf buf) throws DecompressionException {
+        if (unusedInput != null) {
+            throw new IllegalStateException("Not in state NEED_INPUT");
+        }
+        addSomeInput(buf);
+        if (buf.isReadable()) {
+            this.unusedInput = buf;
+        } else {
+            buf.release();
+        }
+    }
+
+    private void addSomeInput(ByteBuf buf) {
+        ByteBuffer decoderInputBuffer = decoder.getInputBuffer();
+        decoderInputBuffer.clear();
+        int readBytes = readBytes(buf, decoderInputBuffer);
+        decoder.push(readBytes);
+    }
+
+    @Override
+    public void endOfInput() throws DecompressionException {
+    }
+
+    @Override
+    public ByteBuf takeOutput() throws DecompressionException {
+        ByteBuffer nativeBuffer = decoder.pull();
+        // nativeBuffer actually wraps brotli's internal buffer so we need to copy its content
+        ByteBuf copy = allocator.buffer(nativeBuffer.remaining());
+        copy.writeBytes(nativeBuffer);
+        return copy;
+    }
+
+    @Override
+    public void close() {
+        decoder.destroy();
+        if (unusedInput != null) {
+            unusedInput.release();
+        }
+    }
+
+    private static int readBytes(ByteBuf in, ByteBuffer dest) {
+        int limit = Math.min(in.readableBytes(), dest.remaining());
+        ByteBuffer slice = dest.slice();
+        slice.limit(limit);
+        in.readBytes(slice);
+        dest.position(dest.position() + limit);
+        return limit;
+    }
+
+    public static Builder builder(ByteBufAllocator allocator) {
+        return new Builder(allocator);
+    }
+
+    public static final class Builder extends AbstractDecompressorBuilder {
+        private int inputBufferSize = 8 * 1024;
+
+        Builder(ByteBufAllocator allocator) {
+            super(allocator);
+        }
+
+        /**
+         * Desired size of the input buffer in bytes. Default 8K.
+         *
+         * @param inputBufferSize desired size of the input buffer in bytes
+         * @return This builder
+         */
+        public Builder inputBufferSize(int inputBufferSize) {
+            this.inputBufferSize = inputBufferSize;
```

Comment:
do we need to validate that this is non-negative ?

Reply:
i'll add a check, I assume otherwise the native code checks this